File: Array&String/GreedyProblem/CPP/partition_array_into_three_parts_with_equal_sum.cpp
# include <numeric>
# include <vector>
using namespace std;

class Solution {
public:
    bool canThreePartsEqualSum(vector<int>& arr) {
        int sums = accumulate(arr.begin(), arr.end(), 0), n = arr.size();
        int mod = sums % 3, rem = sums / 3, parts = 3, cur = 0;
        if (mod) return false;
        for (int i = 0; i < n; i++) {
            cur += arr[i];
            if (parts == 1 && i < n - 1) continue;
            if (cur == rem) {
                cur = 0;
                parts--;
            }
        }
        return parts == 0;
    }

    // Shorter
    bool canThreePartsEqualSum(vector<int>& A) {
        int sum = accumulate(A.begin(), A.end(), 0), part = 0, average = sum / 3, cnt = 0;
        for (int a : A) {
            part += a;
            if (part == average) { // find the average: sum / 3
                ++cnt; // find an average, increase the counter.
                part = 0; // reset part.
            }
        }
        return cnt >= 3 && sum % 3 == 0;
    }
};

File: Array&String/GreedyProblem/Python/partition_array_into_three_parts_with_equal_sum.py
class Solution:
    def canThreePartsEqualSum(self, arr: list[int]) -> bool:
        sums, n = sum(arr), len(arr)
        mod, rem = sums % 3, sums // 3
        if mod: return False
        if not any(arr): return True
        part_n, cur = 3, 0
        for i in range(n):
            cur += arr[i]
            if part_n == 1 and i < n - 1: continue
            if cur == rem:
                cur = 0
                part_n -= 1
        return part_n == 0

    # shorter
    def canThreePartsEqualSum(self, A: list[int]) -> bool:
        average, remainder, part, cnt = sum(A) // 3, sum(A) % 3, 0, 0
        for a in A:
            part += a
            if part == average:
                cnt += 1
                part = 0
        return not remainder and cnt >= 3

    # greedy pointers
    def canThreePartsEqualSum(self, A: list[int]) -> bool:
        l, r, s = 1, len(A) - 2, sum(A)
        if s % 3 != 0:
            return False
        leftSum, rightSum, average = A[0], A[-1], s // 3
        while l <= r:
            if l < r and leftSum != average:
                leftSum += A[l]
                l += 1
            if l < r and rightSum != average:
                rightSum += A[r]
                r -= 1
            if leftSum == average == rightSum:
                return True
            if l == r:
                return False
        return False